### ISSUE TITLE
Winston Logging Transports crashing

### ISSUE THREAD

**Issue by noelbranz** [googleapis/nodejs-logging#381]
#### Environment details

  - OS:  Google App-Engine
  - Node.js version: 8
  - npm version:  6.7.0
  - `@google-cloud/logging` version:  4.2.0
 
#### Steps to reproduce

  1. ?
  2. ?

```
<rejected> TypeError: Cannot read property 'length' of undefined
at BundleExecutor.schedule (/app/node_modules/@google-cloud/logging/node_modules/google-gax/build/src/bundling.js:319:43)
at status.call (/app/node_modules/@google-cloud/logging/node_modules/google-gax/build/src/bundling.js:443:30)
at Canceller.call (/app/node_modules/@google-cloud/logging/node_modules/google-gax/build/src/api_callable.js:78:27)
at Bundleable.call (/app/node_modules/@google-cloud/logging/node_modules/google-gax/build/src/bundling.js:442:20)
at funcWithAuth.then.then.apiCall (/app/node_modules/@google-cloud/logging/node_modules/google-gax/build/src/api_callable.js:312:23)
at <anonymous> at process._tickCallback (internal/process/next_tick.js:189:7)
```

**Comment by JustinBeckwith** [comment]
Greetings!  Can you share the code you're using to get this callstack?

**Comment by noelbranz** [comment]
`const { loggers, format } = require('winston');
const { LoggingWinston } = require('@google-cloud/logging-winston');
const {
  combine, label, printf, splat,
} = format;
const logFormat = printf(info => `[${info.label}] ${info.level.toUpperCase()}: ${info.message}`);
const loggingWinston = new LoggingWinston({
  serviceContext: {
    service: 'api',
    version: pjson.version,
  },
  logName: `logfile-v${pjson.version}.log`,
});
 loggers.add('api', {
    level: 'debug',
    format: combine(label({ label: 'api'}), splat(), logFormat),
    transports: [loggingWinston],
    exitOnError: false,
    silent: false,
  });`

This is how I initialize the logger,  the error just intermittently goes out, cannot locate which actual code.

**Comment by JustinBeckwith** [comment]
@alexander-fenster uh...  do you recognize anything in this call stack?

**Comment by alexander-fenster** [comment]
Somewhere in `bundling.ts` - I will take a deeper look today. Haven't seen this before.

**Comment by alexander-fenster** [comment]
This is where we create `Bundleable` instance:

https://github.com/googleapis/nodejs-logging/blob/05edc862a8d5d052e35497bf5f063a219db880d9/src/v2/logging_service_v2_client.js#L142-L151

Here `bundledField` is `entries`.

The call starts here:

![image](https://user-images.githubusercontent.com/4015807/52587686-68c9cc80-2def-11e9-88b3-48bed0d9cc60.png)

and the sequence of nested calls eventually comes to `BundleExecutor.schedule` where it finds that `request["entries"]` is undefined:

![image](https://user-images.githubusercontent.com/4015807/52588581-ad566780-2df1-11e9-8752-adf9d88aecf5.png)

(and we fail trying to get `.length` of undefined)

Now the fun part is that the Logging library does great job making sure that `entries` property is always defined. By looking at its code:

https://github.com/googleapis/nodejs-logging/blob/6cf18aac0133709770596d405242f86d912fda48/src/log.ts#L783-L807

 I can see the only way of having it undefined: it will happen if `decorateEntries_` throws an exception, which will leave `decoratedEntries` undefined, and `reqOpts.entries` will become undefined as well.

Long story short...  https://github.com/googleapis/gax-nodejs/pull/424 to make GAX not fail in this case, and in scope of this issue, we must make sure that `undefined` is never passed to logging API.

**Comment by stephenplusplus** [comment]
> I can see the only way of having it undefined: it will happen if `decorateEntries_` throws an exception, which will leave `decoratedEntries` undefined, and `reqOpts.entries` will become undefined as well.

It kind of seems like we shouldn't let that be `undefined`. It actually doesn't seem like we should send the request, and we would want to give the error to the user. I think I did [*this* error ignore](https://github.com/googleapis/nodejs-logging/blob/6cf18aac0133709770596d405242f86d912fda48/src/log.ts#L779) because we were just doing our best to get additional information we could send with the request, but it wasn't something the user required us to do. So if it fails, it's not a big deal. But the kind of error that would be thrown in the snippet you shared is when a JSON object cannot be serialized because of circular references. The user would probably want to know that, but maybe my memory is failing me on how this all works.

**Comment by alexander-fenster** [comment]
@stephenplusplus I'm not sure what is our policy re: throwing exceptions from logging calls. I think users don't expect that regular log writes could throw 😱 Would it be OK if we make sure the `decorateEntries_` never throws, and does whatever it can possibly do even for objects with circular references? (e.g. dump incomplete object, or anything else)

**Comment by seriousManual** [comment]
Hello,
I have the same issue on Google AppEngine, with the same error appearing in Error Reporting.

Have you identified any circumstances that lead to this error?
I noticed that a certain (older) version of our application has no issues with this bug. This may relate to other parameters handed to winston (we use the winston transport)? If I knew which conditions are responsible I could prevent this error from happening (for now), until this is fixed.

Would switching to another version help with this issue?

**Comment by alexander-fenster** [comment]
Hello @seriousManual, 

> Have you identified any circumstances that lead to this error?

Yes and no. I made a fix https://github.com/googleapis/gax-nodejs/pull/424 in exactly the place where it fails so it should not fail anymore (we'll release the new version of that package today). When all the dependency chain upgrades, the error condition should stop happening.

As for what causes the error - i.e. what causes the undefined value to be passed to logging - I don't have an answer I'm confident in. One idea is that it might be doing that if it gets an object with circular references, but it makes me think there should be another explanation, given that all these bug reports started coming from several people at the same time. @stephenplusplus can I ask you to take a deeper look at what's going on (why this undefined value appears there)?

**Comment by seriousManual** [comment]
Thank you for the efforts you're putting into this.

> but it makes me think there should be another explanation, given that all these bug reports started coming from several people at the same time.
I'm glad it is not only us.
I am currently trying to pinpoint the time and place in the commit history where a possible change has been introduced that may relate to the issue. But this is quite cumbersome as the only way is to redeploy and wait for the bug to happen / not to happen.

How long do you think it will take until the fix patch has been introduced at the upstream deps? (gax -> nodejs-logging -> winston-logging)

**Comment by noelbranz** [comment]
> Hello,
> I have the same issue on Google AppEngine, with the same error appearing in Error Reporting.
> 
> Have you identified any circumstances that lead to this error?
> I noticed that a certain (older) version of our application has no issues with this bug. This may relate to other parameters handed to winston (we use the winston transport)? If I knew which conditions are responsible I could prevent this error from happening (for now), until this is fixed.
> 
> Would switching to another version help with this issue?

Hi,

I would like to share my workaround as of this time, while waiting for the fix.
I tap into the process.on('unhandledRejection')  and check if it is the same error:  then I use logger.configure(...)   to initialize another LoggingWinston  transport -- (this time another filename, it did worked for me for (although it creates another file in stackdriver, but its ok as I can still capture the logs)

**Comment by noelbranz** [comment]
To all who spent time on this issue, I would like to thank you for your dedication to solve this.  I have been tracing all logging combination that we do in our code (NodeJS in App-Engine) to find what is the exact scenario this happen.  But if this can be solved on your end,  Thank you very much!

**Comment by alexander-fenster** [comment]
Greetings folks,

I have released google-gax v0.25.2 which should not crash in case if an undefined value is passed. From what I understand, it means you may get an incomplete / malformed log message but at least nothing will crash. Since logging uses `"google-gax": "^0.25.0"` the fix will be picked up automatically.

**Comment by seriousManual** [comment]
Hi,
thank you for the fix, I updated the dependencies today, redeployed and promptly got the following output on stderr:
"Request does not contain field entries that must present for bundling. Invoking immediately"

I suppose that this is directly related to the fix? 
Either way, though this case occured the logs keep coming in, so far it's looking good!

**Comment by alexander-fenster** [comment]
@seriousManual Yes, it's exactly that warning message that should be printed in this case. Note that it will be only printed once per each running process (otherwise your `stderr` would've been spammed a lot).

**Comment by noelbranz** [comment]
Hi Guys,  I also did used the fix and deployed to our staging instance.  Its been there for 3 days now and I never got the error again.  Hopefully this is going to be stable now.  Thanks a lot.

**Comment by seriousManual** [comment]
Same here, 
haven't seen the issue on our development system for a while now!

Thank you!

**Comment by la-urre** [comment]
I observe the same issue with winston 3.2.1 and google-gax 1.0.0. I get the same error "Attempt to write logs with no transports" after a while.

I can see there was a huge refactoring done recently, so maybe there was a regression during the process ?

**Comment by ltomes** [comment]
@la-urre Did you ever get to the bottom of your issue? I am seeing the same thing currently.

**Comment by JustinBeckwith** [comment]
If you're running into issues, please open a new bug with details 🤗
